File: C++/5-BinarySearch/Problems/4-KthSmallestElement-in-a-SortedMatrix/KthSmallestElement-in-a-SortedMatrixPro.js
/**
 * Five Approaches to Find the Kth Smallest Element in a Sorted Matrix
 *
 * A sorted matrix is where each row and each column are in ascending order.
 *
 * 1.  Brute Force (Flatten and Sort)
 * 2.  Min-Heap (Priority Queue)
 * 3.  Binary Search
 * 4.  Optimized Binary Search
 * 5.  Iterative Binary Search
 */

const matrix = [
    [1, 5, 9],
    [10, 11, 13],
    [12, 13, 15]
];
const k = 8;

console.log("Input Matrix:");
console.log(matrix);
console.log("K:", k);

// 1. Brute Force (Flatten and Sort)
//    - Time Complexity: O(N*M log(N*M)), where N is the number of rows and M is the number of columns.
//    - Space Complexity: O(N*M)
function kthSmallest_BruteForce(matrix, k) {
    const rows = matrix.length;
    const cols = matrix[0].length;
    const flattened = [];

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            flattened.push(matrix[i][j]);
        }
    }

    flattened.sort((a, b) => a - b); // Sort in ascending order
    return flattened[k - 1]; // k-1 because arrays are zero-indexed
}

console.log("\n1. Brute Force:");
console.log("Kth Smallest Element:", kthSmallest_BruteForce(matrix, k));


// 2. Min-Heap (Priority Queue)
//    - Time Complexity: O(k log k) in best case (k < N), O(N*M log(k)) in worst case (k ~ N*M)
//    - Space Complexity: O(k)
function kthSmallest_MinHeap(matrix, k) {
    const rows = matrix.length;
    const cols = matrix[0].length;
    const minHeap = new MinPriorityQueue(); // Use a min-heap (PriorityQueue)

    // Push the first column elements into the min-heap.
    for (let i = 0; i < rows; i++) {
        minHeap.enqueue({ val: matrix[i][0], row: i, col: 0 }, matrix[i][0]); // {val, row, col}
    }

    let count = 0;
    let result = 0;

    while (count < k) {
        const { val, row, col } = minHeap.dequeue().element;  // Get the smallest
        result = val;
        count++;

        if (col < cols - 1) {
            // Push the next element in the same row
            minHeap.enqueue({ val: matrix[row][col + 1], row: row, col: col + 1 }, matrix[row][col+1]);
        }
    }
    return result;
}

// Implementation of MinPriorityQueue (Simplified for this problem)
class MinPriorityQueue {
  constructor() {
    this.heap = [];
  }

  enqueue(element, priority) {
    this.heap.push({ element, priority });
    this.bubbleUp(this.heap.length - 1);
  }

  dequeue() {
    if (this.isEmpty()) return null;
    this.swap(0, this.heap.length - 1);
    const popped = this.heap.pop();
    this.bubbleDown(0);
    return popped;
  }

  isEmpty() {
    return this.heap.length === 0;
  }

  bubbleUp(index) {
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      if (this.heap[index].priority >= this.heap[parentIndex].priority) break;
      this.swap(index, parentIndex);
      index = parentIndex;
    }
  }

  bubbleDown(index) {
    while (true) {
      const leftChild = 2 * index + 1;
      const rightChild = 2 * index + 2;
      let smallest = index;

      if (leftChild < this.heap.length && this.heap[leftChild].priority < this.heap[smallest].priority) {
        smallest = leftChild;
      }
      if (rightChild < this.heap.length && this.heap[rightChild].priority < this.heap[smallest].priority) {
        smallest = rightChild;
      }
      if (smallest === index) break;
      this.swap(index, smallest);
      index = smallest;
    }
  }

  swap(i, j) {
    [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
  }
}

console.log("\n2. Min-Heap:");
console.log("Kth Smallest Element:", kthSmallest_MinHeap(matrix, k));


// 3. Binary Search
//    - Time Complexity: O(N log(max - min) * log N), where max and min are the largest and smallest elements.
//    - Space Complexity: O(1)
function kthSmallest_BinarySearch(matrix, k) {
    const rows = matrix.length;
    const cols = matrix[0].length;
    let low = matrix[0][0];
    let high = matrix[rows - 1][cols - 1];

    while (low < high) {
        const mid = low + Math.floor((high - low) / 2);
        let count = 0;
        let j = cols - 1;

        for (let i = 0; i < rows; i++) {
            while (j >= 0 && matrix[i][j] > mid) j--;
            count += (j + 1);
        }

        if (count < k) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    return low;
}

console.log("\n3. Binary Search:");
console.log("Kth Smallest Element:", kthSmallest_BinarySearch(matrix, k));



// 4. Optimized Binary Search
//     - Time Complexity: O(N log(max - min)), where max and min are the largest and smallest elements.
//     - Space Complexity: O(1)
function kthSmallest_OptimizedBinarySearch(matrix, k) {
    const n = matrix.length;
    let low = matrix[0][0];
    let high = matrix[n - 1][n - 1];

    while (low < high) {
        const mid = low + Math.floor((high - low) / 2);
        let count = 0;
        let j = n - 1;

        for (let i = 0; i < n; i++) {
            while (j >= 0 && matrix[i][j] > mid) {
                j--;
            }
            count += j + 1;
        }

        if (count < k) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    return low;
}

console.log("\n4. Optimized Binary Search:");
console.log("Kth Smallest Element:", kthSmallest_OptimizedBinarySearch(matrix, k));


// 5. Iterative Binary Search
//    - Time Complexity: O(N log(max - min)), where max and min are the largest and smallest elements.
//    - Space Complexity: O(1)
function kthSmallest_IterativeBinarySearch(matrix, k) {
    const n = matrix.length;
    let left = matrix[0][0];
    let right = matrix[n - 1][n - 1];

    while (left < right) {
        const mid = left + Math.floor((right - left) / 2);
        let count = 0;
        let j = n - 1;

        for (let i = 0; i < n; i++) {
            while (j >= 0 && matrix[i][j] > mid) {
                j--;
            }
            count += j + 1;
        }

        if (count < k) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return left;
}

console.log("\n5. Iterative Binary Search:");
console.log("Kth Smallest Element:", kthSmallest_IterativeBinarySearch(matrix, k));

File: Java/5-BinarySearch/Problems/4-KthSmallestElement-in-a-SortedMatrix/KthSmallestElement-in-a-SortedMatrixPro.java
import java.util.PriorityQueue;
import java.util.Comparator;

public class KthSmallestInSortedMatrix {

    public static void main(String[] args) {
        int[][] matrix = {
                {1, 5, 9},
                {10, 11, 13},
                {12, 13, 15}
        };
        int k = 8;

        // Test all 5 approaches
        System.out.println("Approach 1: Brute Force (Sorting): " + kthSmallest_BruteForce(matrix, k));
        System.out.println("Approach 2: Min Heap: " + kthSmallest_MinHeap(matrix, k));
        System.out.println("Approach 3: Binary Search: " + kthSmallest_BinarySearch(matrix, k));
        System.out.println("Approach 4: Optimized Min Heap: " + kthSmallest_OptimizedMinHeap(matrix, k));
        System.out.println("Approach 5: Iterative Binary Search: " + kthSmallest_IterativeBinarySearch(matrix, k));
    }

    // Approach 1: Brute Force (Sorting)
    // - Time Complexity: O(N^2 log N^2), where N is the dimension of the matrix.
    // - Space Complexity: O(N^2)
    // - Explanation: Flatten the matrix into a 1D array, sort it, and return the k-th element.
    public static int kthSmallest_BruteForce(int[][] matrix, int k) {
        int n = matrix.length;
        int[] flattened = new int[n * n];
        int index = 0;

        // Flatten the matrix
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                flattened[index++] = matrix[i][j];
            }
        }

        // Sort the flattened array
        java.util.Arrays.sort(flattened);

        // Return the k-th smallest element (k is 1-indexed)
        return flattened[k - 1];
    }

    // Approach 2: Min Heap
    // - Time Complexity: O(k log k), where k is the given k.  Can be O(N^2 log k) in the worst case if k is close to N^2.
    // - Space Complexity: O(k)
    // - Explanation: Use a min-heap to store the smallest elements.  Iterate through the matrix,
    //              offer elements to the heap, and if the heap size exceeds k, remove the largest
    //              element in the heap.  The top of the heap will be the k-th smallest.
    public static int kthSmallest_MinHeap(int[][] matrix, int k) {
        int n = matrix.length;
        PriorityQueue<Integer> minHeap = new PriorityQueue<>(Comparator.reverseOrder());

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                minHeap.offer(matrix[i][j]);
                if (minHeap.size() > k) {
                    minHeap.poll();
                }
            }
        }
        return minHeap.peek();
    }

    // Approach 3: Binary Search
    // - Time Complexity: O(N log(max - min)), where N is the dimension of the matrix,
    //                    and max and min are the maximum and minimum elements in the matrix.
    // - Space Complexity: O(1)
    // - Explanation: Perform binary search on the range of values in the matrix. For each mid value,
    //              count how many elements in the matrix are less than or equal to it.  Adjust the
    //              search range based on this count.
    public static int kthSmallest_BinarySearch(int[][] matrix, int k) {
        int n = matrix.length;
        int low = matrix[0][0];
        int high = matrix[n - 1][n - 1];

        while (low < high) {
            int mid = low + (high - low) / 2;
            int count = countElementsLessThanOrEqual(matrix, mid);
            if (count < k) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        return low;
    }

    private static int countElementsLessThanOrEqual(int[][] matrix, int target) {
        int n = matrix.length;
        int row = n - 1;
        int col = 0;
        int count = 0;

        while (row >= 0 && col < n) {
            if (matrix[row][col] <= target) {
                count += row + 1; // Add the entire column up to the current row
                col++;
            } else {
                row--;
            }
        }
        return count;
    }

    // Approach 4: Optimized Min Heap
    // - Time Complexity: O(k log N), where N is the dimension of the matrix.
    // - Space Complexity: O(N)
    // - Explanation:  A min-heap is used, but instead of adding all elements, we start by adding the first row.
    //       Each entry in the heap is an element along with its row and column index.  We repeatedly
    //       extract the minimum, add it to our count, and then add the element below it in the same
    //       column (if it exists) to the heap.  This ensures we only keep at most N elements in the heap.
    public static int kthSmallest_OptimizedMinHeap(int[][] matrix, int k) {
        int n = matrix.length;
        PriorityQueue<Tuple> minHeap = new PriorityQueue<>(Comparator.comparingInt(t -> t.val));

        // Add the first row to the min-heap
        for (int j = 0; j < n; j++) {
            minHeap.offer(new Tuple(matrix[0][j], 0, j));
        }

        int count = 0;
        int result = 0;
        while (count < k) {
            Tuple tuple = minHeap.poll();
            result = tuple.val;
            count++;

            int row = tuple.row;
            int col = tuple.col;
            if (row < n - 1) {
                minHeap.offer(new Tuple(matrix[row + 1][col], row + 1, col));
            }
        }
        return result;
    }

    // Helper class for Optimized Min Heap approach
    private static class Tuple {
        int val;
        int row;
        int col;

        public Tuple(int val, int row, int col) {
            this.val = val;
            this.row = row;
            this.col = col;
        }
    }

    // Approach 5: Iterative Binary Search
    // - Time Complexity: O(N log(max - min)), where N is the dimension of the matrix.
    // - Space Complexity: O(1)
    // - Explanation: This is a non-recursive version of the Binary Search approach.  It uses a while loop
    //   instead of recursion to perform the binary search, but the core logic remains the same.
    public static int kthSmallest_IterativeBinarySearch(int[][] matrix, int k) {
        int n = matrix.length;
        int low = matrix[0][0];
        int high = matrix[n - 1][n - 1];

        while (low <= high) { // Changed condition to low <= high
            int mid = low + (high - low) / 2;
            int count = 0;
            int j = n - 1;

            for (int i = 0; i < n; i++) {
                while (j >= 0 && matrix[i][j] > mid) {
                    j--;
                }
                count += (j + 1); // Important:  j could be -1.
            }

            if (count < k) {
                low = mid + 1;
            } else {
                high = mid - 1; // Changed to high = mid - 1
            }
        }
        return low; // Changed return value to low
    }
}

